### DISCUSSION TITLE
WebRAVE Symbology JSON id Values

### DISCUSSION THREAD

**Discussion by MattReimer** [Riverscapes/RiverscapesXML#109] [RAVE: Symbology/BusinessLogic] [answered]
Moved from: https://github.com/Riverscapes/rs-web-monorepo/issues/123

> @MattReimer I don't know if this is actually related, but when creating the .json files, I've noticed that the ID and source layer tags have extra characters that mapbox adds on after them. Example:
> 
> ```
> "id": "confining-poly-3a679s",
>        "type": "fill",
>        "source": "composite",
>        "source-layer": "confining_poly-3a679s",
> ```
> 
> For Arc and Q I've used just `"confining_poly"` in those spots... again not sure if this is just how .json works?

**Comment by MattReimer** [comment, marked as answer]
@lauren-herbine  It's not so much how JSON works. It's more how MapBox works. Mapbox is ensuring uniqueness by adding hash characters after the layer and source names.

We **_do_** require both "id" and "source-layer" fields but it's only used in the validation of the symbology, not actually in rendering. 

* `id`: Basically I check that `id` exists and that's all. At render time we're going to assign a new one anyway so I don't reallly care about the values themselves
* `source-layer`:  what I'm looking for is that there aren't any Mapbox-proprietary source layers and also that every symbology file references one (and only one) source layer.


### Good Example:

So let's say you had a symbology file where you wanted to have a polygon with both fill and border properties

It might look something like this when you export it.

``` json
[
    {
        "id": "confinement-ratio-3oldrq",
        "type": "fill",
        "source": "composite",
        "source-layer": "confinement_ratio-3oldrq",
        "layout": {},
        "paint": "blah blah blah whatever mapbox stuff here"    },
    {
        "id": "confinement-ratio-sdf823s",
        "type": "line",
        "source": "composite",
        "source-layer": "confinement_ratio-3oldrq",
        "layout": {},
        "paint": "blah blah blah whatever mapbox stuff here"
    }
]
```

You'll notice that they have different `id` values because each block inside the `{ }` braces is a separate render layer that needs its own id. They both reference the same `source-layer` though. This is allowed and perfectly fine. 

### Bad Example

Now let's see a bad example:


``` json
[
    {
        "id": "confinement-ratio-3oldrq",
        "type": "fill",
        "source": "composite",
        "source-layer": "confinement_ratio-3oldrq",
        "layout": {},
        "paint": "blah blah blah whatever mapbox stuff here"    },
    {
        "id": "confinement-ratio-sdf823s",
        "type": "line",
        "source": "composite",
        "source-layer": "constriction_ratio-23dfs2q",
        "layout": {},
        "paint": "blah blah blah whatever mapbox stuff here"
    }
]
```

Here the same sort of thing except the `fill` layer is coming from  `confinement_ratio-3oldrq` and the `line` layer is coming from `constriction_ratio-23dfs2q`. This is allowed by mapbox but not by our system so if you try to commit this code the validator should throw out an error. 

--------------------------------------------

this is maybe more detail that you need to know just to copy and paste some JSON but I'll put it here so there's some justification for the seemingly arbitrary rules we've imposed on this very new process

**Reply by lauren-herbine** [reply]
@MattReimer Thanks for the explanation! I'll try and make sure the source-layers are referencing the right things.
